[Professional-Procrastinat0r] iP

Professional-Procrastinat0r opened the pull request:
# OuiOuiBaguette
> Bone-jaw! Im Oui Oui Baguette.
> What can I do for you? Oui Oui

Your ~~wannabe~~ French-speaking companion :smirk:

## Features
- Add and delete tasks
- List tasks
- Mark and unmark tasks
- Persistent file storage
- Find task (roadmap)

## Commands
1. list
2. todo `task description`
3. deadline `task description` /by `due date`
4. event `task description` /from `start date` /by `end date`
5. mark `index`
6. unmark `index`
7. delete `index`

## Roadmap
Codebase can be accessed on this [Github repo](https://github.com/Professional-Procrastinat0r/ip).
- [x] Save to disk
- [x] Find tasks
- [ ] Add GUI

## Entry Point
Entry point of the application is in Main.java:
```java
import ouiouibaguette.OuiOuiBaguette;

/**
 * The entry point of the OuiOuiBaguette application.
 * Initializes and runs the OuiOuiBaguette bot.
 */
class Main {

    /**
     * The main method that starts the OuiOuiBaguette application.
     *
     * @param args Command-line arguments (not used).
     */
    public static void main(String[] args) {
        OuiOuiBaguette bot = new OuiOuiBaguette("data");
        bot.run();
    }
}

```

## Contributing Guide
All changes are to be made via raising PRs, which should follow a similar format as [this PR]().

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Professional-Procrastinat0r committed:
Adding initial skeletal version of OuiOuiBaguette

* Rename Duke.java to OuiOuiBaguette.java
* Change class name to OuiOuiBaguette

Professional-Procrastinat0r committed:
Add chatbot echoes back commands

Chatbot reads input from user:
* exits if the command is 'bye'
* echoes back otherwise

Professional-Procrastinat0r committed:
Extract formatting for chatbot response

Each chatbot response requires the message to be formatted,
with horizontal lines and indentation.

Response format is standard,
causing duplication in formatting each response.

Let's isolate response formatting to 2 utility functions,
which take in message(s) as
a string or an array of strings.

Using functions allows us to reuse formatting logic,
reducing code duplication.

Professional-Procrastinat0r committed:
Add add task and list task functionality

list command displays list of tasks.

All other input (except bye) adds to list of tasks.

Professional-Procrastinat0r committed:
Close scanner

Scanner is opened but not closed.

Close scanner to prevent resource leak.

Professional-Procrastinat0r committed:
Add marking and unmarking tasks functionality

Add Task class for encapsulating task-related functionality

Professional-Procrastinat0r committed:
Rename Task methods for brevity

Professional-Procrastinat0r committed:
Add todo

Professional-Procrastinat0r committed:
Add deadline

Professional-Procrastinat0r committed:
Add event

Professional-Procrastinat0r committed:
Update main file name

Professional-Procrastinat0r committed:
Add test cases

Test:
* Add 3 task types
* Mark 3 task types
* Unmark 3 task types
* List

Professional-Procrastinat0r committed:
Add base exception class for chatbot

Gives personality to error message

Professional-Procrastinat0r committed:
Add unknown command exception

Thrown when an unknown command is input by the user

Professional-Procrastinat0r committed:
Add exception classes for tasks

Exception classes for tasks implemented as subclasses of
OuiOuiBaguette exception class.

This formats the error message with the
personality of the chatbot.

Professional-Procrastinat0r committed:
Add error-handling to chatbot

Professional-Procrastinat0r committed:
Add delete task functionality

Professional-Procrastinat0r committed:
Add file save data format

Professional-Procrastinat0r committed:
Add read/write tasks to file

Professional-Procrastinat0r committed:
Update gitignore to ignore data files

Professional-Procrastinat0r committed:
Merge branch 'branch-Level-7'

Professional-Procrastinat0r committed:
Format date displayed in CLI

Professional-Procrastinat0r committed:
Update test cases

Professional-Procrastinat0r committed:
Merge branch 'branch-Level-8'

Professional-Procrastinat0r committed:
Add Storage classes

Chatbot writes to file storage directly.

This results in poor abstraction and encapsulation,
as it makes the main driver code responsible for directly
accessing the file storage.

Extracting storage read/writes allow us to
isolate all storage operations.
We can also use polymorphism to replace the FileStorage class
with another storage medium in the future.

Let's put all storage operations in an abstract
class called Storage, and create a subclass FileStorage class
which uses files as storage.

Professional-Procrastinat0r committed:
Add TaskList class

Chatbot writes manages task list and updating file storage directly.

This results in poor abstraction and encapsulation,
as it makes the main driver code responsible for
ensuring that task list and file storage are synchronised.

Creating a TaskList class allows us to update the storage
each time the task list and its elements are modified.

Let's make the TaskList class responsible for updating
the storage.

This enforces better abstraction, as the driver code
does not need to be concerned with managing storage.

Professional-Procrastinat0r committed:
Move task classes into tasks package

Professional-Procrastinat0r committed:
Add UI class

Chatbot manages formatting console messages directly.

This breaks abstraction and encapsulation,
as it makes the main driver code responsible for directly
managing the UI.

Extracting storage read/writes allow us to
isolate all UI operations.
We can also use polymorphism to replace the CommandLineUI class
with another UI in the future.

Let's put all UI operations in CommandLineUI class,
which manages formatting the messages to display in the console.

This enforces better abstraction, as the driver code
does not need to be concerned with managing display.

Professional-Procrastinat0r committed:
Add Parser and Command classes

Chatbot manages parsing the user input
and executing the commands directly.

This breaks single responsibility principle,
as it makes the main driver code responsible for managing
parsing and command logic together.

Extracting parsing allows us to update our parsing logic
independently of our commands.

Let's create a Parser class, which will handle parsing user inputs.
The Parser class will then return the appropriate Command to
execute.

This adheres better to the SRP, as one class should
not manage multiple key functionalities.

Professional-Procrastinat0r committed:
Update main code

Professional-Procrastinat0r committed:
Update testing

Professional-Procrastinat0r committed:
Merge branch 'branch-A-MoreOOP'

Professional-Procrastinat0r committed:
Merge branch 'add-gradle-support' of https://github.com/Professional-Procrastinat0r/ip into add-gradle-support

Professional-Procrastinat0r committed:
Update gradle.build main file

Professional-Procrastinat0r committed:
Merge branch 'add-gradle-support'

Professional-Procrastinat0r committed:
Handle incorrect date formats

Professional-Procrastinat0r committed:
Add JUnit test cases

Professional-Procrastinat0r committed:
Update gitignore

Professional-Procrastinat0r committed:
Merge branch 'branch-A-JUnit'

Professional-Procrastinat0r committed:
Add checkstyle

Professional-Procrastinat0r committed:
Fix checkstyle errors (imports/spaces)

Professional-Procrastinat0r committed:
Fix remaining checkstyle errors (spaces)

Professional-Procrastinat0r committed:
Merge branch 'add-gradle-support'

Professional-Procrastinat0r committed:
Add error handling for tasklist index out of bounds

Professional-Procrastinat0r committed:
Change jar base name

Professional-Procrastinat0r committed:
Merge branch 'branch-A-Jar'

Professional-Procrastinat0r committed:
Add find by keyword

Professional-Procrastinat0r committed:
Add javadocs to find command

Professional-Procrastinat0r committed:
Fix checkstyle error for find command

Professional-Procrastinat0r committed:
Add javadocs to command package

Professional-Procrastinat0r committed:
Add javadocs to storage package

Professional-Procrastinat0r committed:
Add javadocs to tasklist package

Professional-Procrastinat0r committed:
Add javadocs to task package

Professional-Procrastinat0r committed:
Add javadocs to parser package

Professional-Procrastinat0r committed:
Add javadocs to ui package

Professional-Procrastinat0r committed:
Add javadocs to ouiouibaguette package and main

Professional-Procrastinat0r committed:
Fix format code to adhere coding style

Professional-Procrastinat0r committed:
Merge branch 'branch-A-JavaDoc'

Professional-Procrastinat0r committed:
Merge branch 'branch-A-CodingStandard'

Professional-Procrastinat0r committed:
Merge branch 'branch-Level-9'

Professional-Procrastinat0r committed:
Change ui class name to match coding standard

Professional-Procrastinat0r committed:
Add javafx to build.gradle

Professional-Procrastinat0r committed:
Remove ui argument for execute method in command

Professional-Procrastinat0r committed:
Add javafx GUI

Professional-Procrastinat0r committed:
Merge branch 'branch-Level-10'

Professional-Procrastinat0r committed:
Modify dialogbox factory methods to take in variable number of dialog strings

Professional-Procrastinat0r committed:
Merge branch 'branch-A-Varargs'

Eru111 commented on src/main/java/ouiouibaguette/OuiOuiBaguette.java:
```diff
@@ -0,0 +1,80 @@
+
+    /**
+     * Generates a response for the user's chat message.
+     */
```
Should there be an @return in the JavaDoc comment? I noticed this in several other places too.

Eru111 commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,140 @@
+    public Command parseCommand(String cmd) throws OuiOuiBaguetteException {
+        if (cmd.equals("bye")) {
+            return new ByeCommand();
+
```
Should this white space be removed? I noticed this in several other places too.

Eru111 commented on src/main/java/storage/FileStorage.java:
```diff
@@ -0,0 +1,117 @@
+            file.createNewFile();
+        } catch (IOException ioException) {
+            System.out.println("Unable to create task save file at path " + file.getAbsolutePath()
+                    + "\n" + ioException.getMessage());
```
I like how you followed the java coding standard for wrapping lines by breaking the line before the '+' operator which I think is a common mistake others might fall for.

Eru111 commented on src/main/java/Main.java:
```diff
@@ -0,0 +1,134 @@
+        // Greetings
+        dialogContainer.getChildren().add(
+            DialogBox.getChatbotDialog(chatbotImage, chatbot.greet())
+        );
```
I like the usage of whitespaces here, makes the method look very clean.

Eru111 reviewed (commented):
I like how clean the code is overall.

Professional-Procrastinat0r committed:
Add assertions to methods in command classes

Subclasses of Command do not have any checks and validations
to input arguments.
For example, the method execute expects a TaskList object
to call list operation methods on.

Null arguments could accidentally be passed, and may result
in NullPointerExceptions if we try to call such methods on the object.

By adding assertions, we check that arguments passed are not null.

This allows us to catch if null arguments are accidentally passed.

Let's add assert statements to check if arguments are not null.

Professional-Procrastinat0r committed:
Add assertions parse method in Parser

Parser class does not have any checks and validations
to input argument.
In this case, the method expects a string command to parse.

Null argument could accidentally be passed, and may result
in NullPointerExceptions if we try to call String methods on it.

By adding assertions, we check that arguments passed are not null.

This allows us to catch if null arguments are accidentally passed.

Let's add assert statements to check if arguments are not null.

Professional-Procrastinat0r committed:
Add assertions to contructor of FileStorage

The constructor does not have any checks and validations
to input argument.
In this case, the constructor expects a path to create the directory.

Null argument could accidentally be passed, and may result
in NullPointerExceptions if we try to create directories or
folders.

By adding assertions, we check that arguments passed are not null.

This allows us to catch if null arguments are accidentally passed.

Let's add assert statements to check if arguments are not null.

Professional-Procrastinat0r committed:
Add assertions to constructor of TaskList

The constructor does not have any checks and validations
to input argument.
In this case, the constructor expects a FileStorage object.

Null argument could accidentally be passed, and may result
in NullPointerExceptions if we try to call operations on the object.

By adding assertions, we check that arguments passed are not null.

This allows us to catch if null arguments are accidentally passed.

Let's add assert statements to check if arguments are not null.

Professional-Procrastinat0r committed:
Add individual methods to generate corresponding Command object

Command string is parsed in a long chain of if-else in
parseCommand method.

This does not follow SLAP, as it makes 1 method responsible
for checking, parsing, and creating the relevant command object
for all commands. This also makes the metho body very long.

Let's isolate parsing of command string and creation of
Command object for each command to its own method.

This adheres more closely to SLAP.

Professional-Procrastinat0r committed:
Use Stream for find

Professional-Procrastinat0r committed:
Merge pull request #1 from Professional-Procrastinat0r/branch-A-Assertions

Branch a assertions

Professional-Procrastinat0r committed:
Merge branch 'master' into branch-A-CodeQuality

Professional-Procrastinat0r committed:
Merge pull request #2 from Professional-Procrastinat0r/branch-A-CodeQuality

Add individual methods to generate corresponding Command object

Professional-Procrastinat0r committed:
Merge branch 'master' into branch-A-Streams

Professional-Procrastinat0r committed:
Merge pull request #3 from Professional-Procrastinat0r/branch-A-Streams

Use Stream for find

Professional-Procrastinat0r committed:
Add Github actions yml file

Professional-Procrastinat0r committed:
Remove unused ui class

Professional-Procrastinat0r committed:
Merge branch 'master' into branch-A-CI

Professional-Procrastinat0r committed:
Add JUnit test and checkstyle to yml file

Professional-Procrastinat0r committed:
Merge pull request #4 from Professional-Procrastinat0r/branch-A-CI

Branch a ci

ItsMeNengYi commented on src/main/java/tasklist/TaskList.java:
```diff
@@ -0,0 +1,193 @@
+    protected Storage storage;
+
+    /** The list that holds all the tasks. */
+    protected List<Task> tasklist;
```
Consider renaming tasklist to tasks for simplicity and clarity, as the List type already implies it’s a collection.

ItsMeNengYi commented on src/main/java/tasklist/TaskList.java:
```diff
@@ -0,0 +1,193 @@
+        // Load data from storage
+        List<String> data = storage.load();
+
+        for (String line : data) {
```
Would it be better rename line to taskEntry? This might improve clarity by making the purpose of the variable more evident.

ItsMeNengYi reviewed (commented):
Great job on OOP and following most of the coding conventions. Just some suggestions in the comments below.

Professional-Procrastinat0r committed:
Add print error message to console

Professional-Procrastinat0r committed:
Extract parse task method

Professional-Procrastinat0r committed:
Remove dead code

Professional-Procrastinat0r committed:
Improve dialogbox styling

Professional-Procrastinat0r committed:
Merge branch 'branch-A-BetterGui'

Professional-Procrastinat0r committed:
Add representative screenshot

Professional-Procrastinat0r committed:
Add acknowledgement to CHatGPT

Professional-Procrastinat0r committed:
Update image assets and product screenshot

Professional-Procrastinat0r committed:
Update readme

Professional-Procrastinat0r committed:
Merge branch 'branch-A-UserGuide'

Professional-Procrastinat0r committed:
Fix readme formatting

Professional-Procrastinat0r committed:
Fix readme image url

Professional-Procrastinat0r committed:
Fix readme image url

Professional-Procrastinat0r committed:
Fix readme image url

Professional-Procrastinat0r committed:
Merge branch 'master' of https://github.com/Professional-Procrastinat0r/ip

Professional-Procrastinat0r committed:
Fix readme table formatting

Professional-Procrastinat0r committed:
Fix readme table formatting

Professional-Procrastinat0r committed:
Move unknown command exception

Professional-Procrastinat0r committed:
Handle invalid index
